### PR TITLE
feat: Implement Sink Attention

### PR DESCRIPTION
**Pull request by aoxy**
Fixes https://github.com/Dao-AILab/flash-attention/issues/1802

## Description

This pull request introduces support for the Sink Attention mechanism (directly following the implementation in GPT-OSS).

This is implemented by optionally incorporating a new `learnable_sink` parameter into the attention functions. This parameter is a tensor of shape `(nheads,)`, providing a learnable bias for each attention head. The sink value is added as an extra logit in the attention score calculation before softmax, allowing the model to learn to "sink" a portion of the attention to a global entry, enhancing model capacity when handling long sequences.

### Key Changes:

- **Python Interface (`flash_attn_interface.py`):**  
  Added the `learnable_sink` parameter to all major forward functions (`flash_attn_func`, `flash_attn_qkvpacked_func`, etc.), including their variable-length counterparts. The backward pass is updated to compute and return the gradient for `learnable_sink`.
- **C++/CUDA Kernels (`csrc/`):**  
  The C++ API and CUDA kernels now accept and integrate the sink values, incorporating them into the softmax computation during the forward pass, and calculating their gradients during the backward pass. A new template parameter `Has_sink` is introduced to conditionally compile the sink-related logic.
- **Testing (`tests/test_flash_attn.py`):**  
  Comprehensive tests, including new test cases with `has_learnable_sink=True`, have been added to validate both forward and backward passes. The numerical correctness of outputs and gradients is verified against reference implementations. All tests pass.
- **Benchmarks (`benchmarks/benchmark_flash_attention.py`):**  
  The benchmark script is updated to add a "Flash2Sink" method for measuring the feature's performance impact.

### CODE REVIEW

**Comment by aoxy** [comment]
My current implementation approach is to add a new top-level interface flash_attn_sink_func, which results in a lot of redundant code. Would you consider adding the sink parameter to the existing interfaces instead?

**Comment by tridao** [comment]
It's better to add to existing interface instead of duplicating code

**Comment by aoxy** [comment]
Hi @tridao , I have updated the PR addressing your feedback—please kindly take another look when you have time. Thanks again for your guidance!

**Comment by aoxy** [comment]
I plan to add sink to the hopper version in the next PR.

File: flash_attn/flash_attn_interface.py [hunk 1]
```diff
@@ -1136,6 +1197,7 @@ def flash_attn_func(
     q,
     k,
     v,
+    learnable_sink,
```
Comment (vasqu) [inline flash_attn/flash_attn_interface.py]: Seems like most interfaces take it as arg? Can we make it optional like the cute version, see e.g. https://github.com/Dao-AILab/flash-attention/blob/c00f8064861936463e5d01648532e10fc7482b9d/flash_attn/cute/interface.py#L518

Comment (aoxy) [reply flash_attn/flash_attn_interface.py]: Done. I made it optional.

**Comment by gunjunlee** [comment]
It seems `flash_attn_with_kvcache` with sink produces incorrect results during decoding. Could you check it?

File: csrc/flash_attn/src/flash_fwd_kernel.h [hunk 2]
```diff
@@ -994,7 +997,8 @@ inline __device__ void compute_attn_1rowblock_splitkv(const Params &params, cons
 
     // Epilogue
 
-    Tensor lse = softmax.template normalize_softmax_lse</*Is_dropout=*/false, Split>(acc_o, params.scale_softmax);
+    // We fix the lse for sink in combine_attn_seqk_parallel.
+    Tensor lse = softmax.template normalize_softmax_lse</*Is_dropout=*/false, Split, /*Has_sink=*/false>(acc_o, params.scale_softmax);
```
Comment (gunjunlee) [inline csrc/flash_attn/src/flash_fwd_kernel.h]: Could you check if `Tensor lse = softmax.template normalize_softmax_lse</*Is_dropout=*/false, Split, /*Has_sink=*/!Split && Has_sink>(acc_o, params.scale_softmax);` is correct?

File: csrc/flash_attn/src/flash_fwd_kernel.h [hunk 3]
```diff
@@ -1187,15 +1191,41 @@ inline __device__ void combine_attn_seqk_parallel(const Params &params) {
...
     lse_sum = Allreduce<kRowsPerLoadTranspose>::run(lse_sum, sum_op);
     // For the case where all local lse == -INFINITY, we want to set lse_logsum to INFINITY. Otherwise
     // lse_logsum is log(0.0) = -INFINITY and we get NaN when we do lse_accum(l) - lse_logsum.
-    ElementAccum lse_logsum = (lse_sum == 0.f || lse_sum != lse_sum) ? INFINITY : logf(lse_sum) + lse_max;
+    ElementAccum lse_logsum = (lse_sum == 0.f || lse_sum != lse_sum) ? INFINITY : __logf(lse_sum) + lse_max;
     // if (bidx == 0 && tidx < 32) { printf("tidx = %d, lse = %f, lse_max = %f, lse_logsum = %f\n", tidx, lse_accum(0), lse_max, lse_logsum); }
+
+    if constexpr (Has_sink) {
+        #pragma unroll
+        for (int l = 0; l < kNLsePerThread; ++l) {
```
Comment (muoshuosha) [inline csrc/flash_attn/src/flash_fwd_kernel.h]: If kNLsePerThread > 1，the attention sink will be repeatedly applied across all num_splits LSE. Is that correct?

Comment (muoshuosha) [reply csrc/flash_attn/src/flash_fwd_kernel.h]: I’ve encountered some precision issues here In case which num_splits is very large(like 128).

**Comment by guilhermeleobas** [comment]
Hi @aoxy, will this work be also ported to work on flash attention 3 (the code on hopper/ subdir)?

**Comment by aoxy** [comment]
@guilhermeleobas , Yes, I also plan to port this work to FlashAttention-3.

File: csrc/flash_attn/src/flash_fwd_kernel.h (hunk 3 above)
Comment (aoxy) [reply csrc/flash_attn/src/flash_fwd_kernel.h]: Fixed.

File: csrc/flash_attn/src/flash_fwd_kernel.h (hunk 2 above)
Comment (aoxy) [reply csrc/flash_attn/src/flash_fwd_kernel.h]: Updated

**Comment by aoxy** [comment]
Hi @tridao ,

Could you please help review this PR?

**All tests pass.** Would appreciate your feedback. Thank you!

**Comment by Potatooff** [comment]
Please

**Comment by aoxy** [comment]
Hi @tridao ,

Sorry to disturb, but may I kindly ask if there are any updates regarding the review of this PR?

Also, I wonder if you are still considering the integration of Sink Attention into Flash Attention v2.

Thank you very much!

**Comment by liuqianchao** [comment]
@aoxy hi, any update on the merge work? 

We’ve recently run into low training efficiency when doing RL training with gpt-oss because sink attention support is inconsistent between training and inference, and we’d like to know when FA2/FA3 are expected to officially support sink attention. 

For more information, you can have a look at https://github.com/volcengine/verl/issues/3794

**Comment by aoxy** [comment]
> @aoxy hi, any update on the merge work?
> 
> We’ve recently run into low training efficiency when doing RL training with gpt-oss because sink attention support is inconsistent between training and inference, and we’d like to know when FA2/FA3 are expected to officially support sink attention.
> 
> For more information, you can have a look at [volcengine/verl#3794](https://github.com/volcengine/verl/issues/3794)

I don't know, you could consider using an internal version.

**Comment by EthanAndersonUSA** [comment]
@tridao any chance for the review of that PR? this would enable using GPT-OSS on some backends using flash-attention like Exllamav3+TabbyAPI

**Comment by jordane95** [comment]
Hi @tridao  

Would you mind taking time to review this PR? This would be very valuable for the community.

Thanks.

**Comment by yadirhb** [comment]
Any updates here?

**Comment by iFe1er** [comment]
@tridao Hi, would you mind taking time to review this PR? This is very important to the community and anyone who need to finetune GPT-OSS series models.

**Comment by iFe1er** [comment]
> > @aoxy hi, any update on the merge work?@aoxy 你好，合并工作有什么更新吗？
> > We’ve recently run into low training efficiency when doing RL training with gpt-oss because sink attention support is inconsistent between training and inference, and we’d like to know when FA2/FA3 are expected to officially support sink attention.最近，我们在使用 gpt-oss 进行 RL 训练时遇到了训练效率低下的问题，因为训练和推理之间的 Sink Attention 支持不一致，我们想知道 FA2/FA3 预计何时正式支持 Sink Attention。
> > For more information, you can have a look at [volcengine/verl#3794](https://github.com/volcengine/verl/issues/3794)欲了解更多信息，您可以查看 volcengine/verl#3794
> 
> I don't know, you could consider using an internal version.我不知道，你可以考虑使用内部版本。

@aoxy hi, I have tried to install FA2 from your feature/attention_with_sink branch of the fork https://github.com/aoxy/flash-attention.git. Then I tested to train GPT-OSS-20B with VERL using your modified version of FA2 (by passing attention_impl parameter). I run it in a totally new envirment so there is no FA2 conflicts.

However, looking at the training metrics, Sink Attention implementation seems still not working. Specifically, during warmup steps, rollout_ppl=2.88 while training_ppl is over 1e7! This bizarre training PPL looks basically the same as my previous experiment using official FA2 (flash_attn-2.8.1, which doesn't support sink). 

Am i doing something wrong or am i missing something? Any suggesion will be appreciated. To discuss more details, you can also contact my RTX: frankyliang. Thanks in advance!

**Comment by jiosephlee** [comment]
@vasqu Hey Anton, I don't want to bother you further but if you have time to take a look at this, that would be awesome. Gpt-oss having FA2 support would be huge for fine-tuning settings that can benefit from both (1) faster bwd and (2) padding-free packing (w/ no cross contamination) that only FA currently supports out of the box
